Mesh_3/include/CGAL/Labeled_mesh_domain_3.h

```diff
@@ -126,15 +126,31 @@ namespace internal {
     }
   };
 
-  template<typename Mesh_domain, typename DetectFunctor>
+  template<typename MeshDomain, typename DetectFunctor>
+  struct Detect_features_in_domain {
+    void operator()(const CGAL::Image_3& image,
+                    MeshDomain& domain,
+                    DetectFunctor functor) const {
+      return functor(image, domain);
+    }
+  };
+  // specialization for `Null_functor`: create the default functor
+  template<typename MeshDomain>
+  struct Detect_features_in_domain<MeshDomain, Null_functor> {
+    void operator()(const CGAL::Image_3& image,
+                    MeshDomain& domain,
+                    Null_functor) const {
+      return;
+    }
+  };
+
+  template<typename MeshDomain, typename DetectFunctor>
   void detect_features(const CGAL::Image_3& image,
-                       Mesh_domain& domain,
+                       MeshDomain& domain,
                        DetectFunctor functor)
   {
-    if (boost::is_same<DetectFunctor, Null_functor>::value)
```

Comment:
Note that with C++17, you could have wrote:
```c++
if constexpr (boost::is_same<DetectFunctor, Null_functor>::value)
```

and avoid the dance with the partial specialization.